### PR TITLE
CORE-18879: Ensure verifying notary implementation

### PR DESCRIPTION
**Pull request by jennyang-r3**
In this PR, I added more unit test cases for verifying notary, and checking any missing checks, and refactored the test.

**added tests:**
- Contract verifying notary should successfully notarise if a signed tx verifies signatory signatures
- Contract verifying notary should respond with error if contract verification failed
- Contract verifying notary should respond with error if a signed transaction failed to signatory signature verification
- Contract verifying notary should respond with error if either current or filtered tx is invalid 
- Contract verifying notary should respond with error if dependency is invalid
- Contract verifying notary should respond with error if dependency reference is missing
- Contract verifying notary should respond with error if dependency input is missing

### CODE REVIEW

File: notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/main/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImpl.kt [hunk 1]
```diff
@@ -196,6 +196,10 @@ class ContractVerifyingNotaryServerFlowImpl() : ResponderFlow {
         try {
 
             val dependentStateAndRefs = filteredTransactionsAndSignatures.flatMap { (filteredTransaction, _) ->
+                require(filteredTransaction.id != initialTransaction.id) {
+                    "Either filtered transaction \"${filteredTransaction.id}\" or " +
+                            "current transaction \"${initialTransaction.id}\" is invalid."
+                }
```
Comment (blsemo) [inline notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/main/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImpl.kt]: Why do we need this check? And if we need this check, shouldn't the error message be more specific to what the problem is, i.e. a transaction depending on itself?

File: notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt [hunk 2]
```diff
@@ -299,8 +401,13 @@ class ContractVerifyingNotaryServerFlowImplTest {
     }
 
     @Test
+    @Order(10)
     fun `Contract verifying notary plugin server should respond with error if transaction verification fails`() {
```
Comment (blsemo) [inline notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt]: That title is wrong - it suggests that the transaction verification (i.e. of the  initital transaction) fails - but it really is the Merkle proof verification of the dependencies that fails here.

File: notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt [hunk 3]
```diff
@@ -351,99 +463,278 @@ class ContractVerifyingNotaryServerFlowImplTest {
...
+        val responseError = responseFromServer.first().error
+        assertThat(responseError).isNotNull
+        assertThat(responseError).isInstanceOf(NotaryExceptionTransactionVerificationFailure::class.java)
+        assertThat((responseError as NotaryExceptionTransactionVerificationFailure).errorText)
+            .contains("Missing reference state and ref from the filtered transaction")
+    }
+
+    @Test
+    @Order(16)
+    fun `Contract verifying notary should respond with error if either current or filtered tx is invalid`() {
```
Comment (blsemo) [inline notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt]: I don't understand the test case - in the real world, the initial transaction cannot be a dependency of itself, as it's not a valid, notarized transaction at this point yet.
If anything, this should complain that there is a circular dependency.

File: notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt [hunk 4]
```diff
@@ -351,99 +463,278 @@ class ContractVerifyingNotaryServerFlowImplTest {
...
+        val responseError = responseFromServer.first().error
+        assertThat(responseError).isNotNull
+        assertThat(responseError).isInstanceOf(NotaryExceptionTransactionVerificationFailure::class.java)
+        assertThat((responseError as NotaryExceptionTransactionVerificationFailure).errorText)
+            .contains("Either filtered transaction \"$SIGNED_TX_ID\" or current transaction \"$SIGNED_TX_ID\" is invalid")
+    }
+
+    @Test
+    @Order(17)
+    fun `Contract verifying notary should respond with error if a signed transaction failed to signatory signature verification`() {
```
Comment (blsemo) [inline notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt]:
```suggestion
    fun `Contract verifying notary should respond with error if a signed transaction failed to verify signatory signature verification`() {
```

File: notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt [hunk 5]
```diff
@@ -351,99 +463,278 @@ class ContractVerifyingNotaryServerFlowImplTest {
...
+        )
+        val responseError = responseFromServer.first().error
+        assertThat(responseError).isNotNull
+        assertThat(responseError).isInstanceOf(NotaryExceptionGeneral::class.java)
+        assertThat((responseError as NotaryExceptionGeneral).message)
+            .contains("Signature failed error")
+    }
+
+    @Test
+    @Order(Integer.MAX_VALUE)
```
Comment (blsemo) [inline notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt]: Why does this test need to be run last?

File: notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt [hunk 6]
```diff
@@ -351,99 +463,278 @@ class ContractVerifyingNotaryServerFlowImplTest {
...
+    @Order(19)
+    fun `Contract verifying notary should successfully notarise if a signed tx verifies signatory signatures`() {
+        whenever(mockTransactionSignatureService.signBatch(any(), any())).thenReturn(
+            listOf(listOf(signatorySignatureCharlie))
+        )
+
+        val signedTransaction = mockSignedTransaction()
+        whenever(signedTransaction.verifySignatorySignatures()).thenAnswer {  }
+
+        signedTransaction.verifySignatorySignatures()
```
Comment (blsemo) [inline notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt]: What is the purpose of this call?

**Review by blsemo — COMMENTED** [review]
I have few questions/remarks about the tests in this PR, see the comments inline, also:
- Why is the ordering of the tests required? Does it make a difference in which order the tests are run? If yes, that is a bad code smell, if no, can we get rid of it?
- A lot of these tests just check that exceptions thrown by other services are converted correctly - they do not check that the real thing would behave as expected on wrong or malicious inputs, so I think we still need to test most of these things as e2e tests. This is especially true for any test where you set a mock API to throw.

File: notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt (hunk 6 above)
Comment (jennyang-r3) [reply notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt]: I was trying to call the actual function, but it didn't work so can be ignored.

File: notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt (hunk 5 above)
Comment (jennyang-r3) [reply notary-plugins/notary-plugin-contract-verifying/notary-plugin-contract-verifying-server/src/test/kotlin/com/r3/corda/notary/plugin/contractverifying/server/ContractVerifyingNotaryServerFlowImplTest.kt]: Ordered test sorted out

**Review by blsemo — APPROVED** [review]
LGTM

**Pull request merged**